paddle/framework/framework.md

```diff
@@ -0,0 +1,205 @@
+
+Each operator has a corresponding gradient operator that defines the gradient computation.
+
+### Layers
+
+If we describe a neural network by calling operator creators directly, our code would be lengthy.  It is easier to use layers creators, which, in addition to calling operator creators for the computation, creates model parameters.
+
+For example, `paddle.layer.fc` and `paddle.layer.conv` both call `paddle.op.gemm`. They should also create and initialize `W`, the layer's parameter, and `act` and `scale`, the attributes.
+
+### Variables
+
+We prefer to represent inputs, outputs, and attributes in Variables, so an operation's output can be used to feed and to configure  other operations.
+
+Variables should be able to hold the following types of values:
+
+- int
+- bool
+- float
+- half
+- string
+- tensor
+- operator
+- scope
+
+Some Python API proposals:
+
+```python
+x = paddle.variable.new("x") # create a variable of not yet known type
+x = paddle.variable.tensor("x") # create a tensor typed varible without value
+x = paddle.variable.int("x", 10) # create an unnamed int varaible and set to 10
+x = paddle.variable.operation("x", paddle.operator.gemm(...)) # an operation
+x = paddle.variable.tensor("x", 
+      numpy.random.randn(200, 100), # set the value of the tensor
+      estimated = true)             # will be updated by the backward algorithm.
+x.estimated = false                 # prevents from update.
+```
+
+Note that the variable has the following methods:
+
+```cpp
+class Variable {
+ public:
+  bool Estimated() const;
+  bool SetEstimated(bool);
+  template <typename T> const T& Get() const;
+  template <typename T> T* GetMutable();
+  template <typename T> bool IsType(T) const;
+};
+```
+
+`Get` and `GetMutable` implements *lazy memory allocation*, as described in the [Variable design doc](https://github.com/PaddlePaddle/Paddle/blob/develop/paddle/framework/variable.md).
+
+Note that `name` is not a property of a variable.  A variable can have various names in different scopes.
+
+### Scope
+
+In programming languages, variables belong to scopes.  Scopes enable the release of local variables when a stack frame pops.
+
+A neural network is arguably equivalent to a program.  The *recurrent operator*, `paddle::operator::Recurrent`, is like `for` loop, and the conditional operator, `paddle::operator::Conditional`, is like `if/switch`.  They can have sub-networks as attributes, and `paddle::operator::Recurrent/Conditional::Run` creates a local scope before executing the sub-network.  At inference time, once `paddle::operator::Recurrent/Conditional::Run` frees the scope before completes.  At training time, it is the corresponding gradient operations' `paddle::operator::RecurrentGrad/ConditionalGrad::Run` who free the local scope.
+
+The global and nested local scopes form a hierarchy.  The following Python functions make it convenient to program scopes:
+
+1. `paddle.scope.current()` returns the current scope, which defaults to
```

Comment:
I meant that it defaults to the value returned by the next (second) line. I will rewrite it to make it clear.